// @react-pdf/primitives/lib/index.js
"use strict";

exports.__esModule = true;
exports.View = exports.Tspan = exports.TextInstance = exports.Text = exports.Svg = exports.Stop = exports.Rect = exports.RadialGradient = exports.Polyline = exports.Polygon = exports.Path = exports.Page = exports.Note = exports.Link = exports.LinearGradient = exports.Line = exports.Image = exports.G = exports.Ellipse = exports.Document = exports.Defs = exports.ClipPath = exports.Circle = exports.Canvas = void 0;
var G = 'G';
exports.G = G;
var Svg = 'SVG';
exports.Svg = Svg;
var View = 'VIEW';
exports.View = View;
var Text = 'TEXT';
exports.Text = Text;
var Link = 'LINK';
exports.Link = Link;
var Page = 'PAGE';
exports.Page = Page;
var Note = 'NOTE';
exports.Note = Note;
var Path = 'PATH';
exports.Path = Path;
var Rect = 'RECT';
exports.Rect = Rect;
var Line = 'LINE';
exports.Line = Line;
var Stop = 'STOP';
exports.Stop = Stop;
var Defs = 'DEFS';
exports.Defs = Defs;
var Image = 'IMAGE';
exports.Image = Image;
var Tspan = 'TSPAN';
exports.Tspan = Tspan;
var Canvas = 'CANVAS';
exports.Canvas = Canvas;
var Circle = 'CIRCLE';
exports.Circle = Circle;
var Ellipse = 'ELLIPSE';
exports.Ellipse = Ellipse;
var Polygon = 'POLYGON';
exports.Polygon = Polygon;
var Document = 'DOCUMENT';
exports.Document = Document;
var Polyline = 'POLYLINE';
exports.Polyline = Polyline;
var ClipPath = 'CLIP_PATH';
exports.ClipPath = ClipPath;
var TextInstance = 'TEXT_INSTANCE';
exports.TextInstance = TextInstance;
var LinearGradient = 'LINEAR_GRADIENT';
exports.LinearGradient = LinearGradient;
var RadialGradient = 'RADIAL_GRADIENT';
exports.RadialGradient = RadialGradient;
